### PR TITLE
[DEV-2149] add transaction table

### PR DESCRIPTION
**Pull request by Yannick1712**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations (on [DFX services](https://dev.services.dfx.swiss))
  - Login/logout
  - Buy/sell payment request
  - KYC page

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

**Comment by Yannick1712** [comment]
Add migration

File: src/subdomains/core/transaction/transaction.service.ts
Comment (davidleomay) [inline src/subdomains/core/transaction/transaction.service.ts:1]: Would it make sense to move this to `/subdomains/supporting/payment`?

**Comment by Yannick1712** [comment]
StakingRewards, RefRewards? Not implemented right now

File: src/subdomains/supporting/payment/services/transaction-job.service.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    if (DisabledProcess(Process.SYNCHRONIZE_TRANSACTION)) return;
+
+    const sortedUnassignedTx = Util.sort(
+      [
+        ...(await this.bankTxService.getBankTxWithoutTransaction()),
+        ...(await this.payInService.getCryptoInputWithoutTransaction()),
+        ...(await this.checkoutTxService.getCheckoutTxWithoutTransaction()),
+      ],
+      'created',
+    );
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/transaction-job.service.ts:40]: RefRewards missing

File: src/subdomains/supporting/payment/entities/transaction.entity.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+export enum TransactionType {
+  BUY_CRYPTO = 'BuyCrypto',
+  BUY_FIAT = 'BuyFiat',
+  INTERNAL = 'Internal',
+  BANK_TX_RETURN = 'BankTxReturn',
+  BANK_TX_REPEAT = 'BankTxRepeat',
+  BUY_CRYPTO_RETURN = 'BuyCryptoReturn',
+  STAKING = 'Staking',
+}
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/entities/transaction.entity.ts:19]: Should we add a CryptoCrypto type?

File: src/subdomains/supporting/payment/entities/transaction.entity.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+export enum TransactionSourceType {
+  BANK_TX = 'BankTx',
+  CRYPTO_INPUT = 'CryptoInput',
+  CHECKOUT_TX = 'CheckoutTx',
+}
+
+@Entity()
+export class Transaction extends IEntity {
+  @Column({ type: 'integer' })
+  sourceId: number;
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/entities/transaction.entity.ts:30]: Can be removed

File: src/subdomains/core/buy-crypto/process/services/buy-crypto.service.ts [hunk 4]
```diff
@@ -55,12 +57,19 @@ export class BuyCryptoService {
...
 
   async createFromBankTx(bankTx: BankTx, buyId: number): Promise<void> {
     let entity = await this.buyCryptoRepo.findOneBy({ bankTx: { id: bankTx.id } });
     if (entity) throw new ConflictException('There is already a buy-crypto for the specified bank TX');
 
+    const transaction = !DisabledProcess(Process.CREATE_TRANSACTION)
+      ? await this.transactionService.update(bankTx.id, TransactionSourceType.BANK_TX, {
+          type: TransactionType.BUY_CRYPTO,
+        })
+      : null;
```
Comment (davidleomay) [inline src/subdomains/core/buy-crypto/process/services/buy-crypto.service.ts:71]: Update with transaction ID?
`.update(bankTx.transaction.id, ...)`?

File: src/subdomains/supporting/fiat-payin/services/fiat-payin-sync.service.ts [hunk 5]
```diff
@@ -32,6 +35,11 @@ export class FiatPayInSyncService {
     for (const payment of payments) {
       try {
         const checkoutTx = await this.createCheckoutTx(payment);
+        if (!DisabledProcess(Process.CREATE_TRANSACTION))
+          await this.transactionService.create({
+            sourceId: checkoutTx.id,
+            sourceType: TransactionSourceType.CHECKOUT_TX,
+          });
```
Comment (davidleomay) [inline src/subdomains/supporting/fiat-payin/services/fiat-payin-sync.service.ts:42]: Save transaction to checkout TX?

File: src/subdomains/supporting/payin/strategies/register/impl/base/register.strategy.ts [hunk 6]
```diff
@@ -49,7 +53,14 @@ export abstract class RegisterStrategy implements OnModuleInit, OnModuleDestroy
...
     const payIns = transactions.map((t) => this.payInFactory.createFromEntry(t));
 
-    await this.payInRepository.saveMany(payIns);
+    for (const payIn of payIns) {
+      const cryptoInput = await this.payInRepository.save(payIn);
+      if (!DisabledProcess(Process.CREATE_TRANSACTION))
+        await this.transactionService.create({
+          sourceId: cryptoInput.id,
+          sourceType: TransactionSourceType.CRYPTO_INPUT,
+        });
```
Comment (davidleomay) [inline src/subdomains/supporting/payin/strategies/register/impl/base/register.strategy.ts:62]: Save transaction to crypto input?

File: src/subdomains/supporting/bank-tx/bank-tx/bank-tx.service.ts [hunk 7]
```diff
@@ -73,9 +99,11 @@ export class BankTxService {
     );
     const allTransactions = olkyTransactions.concat(revolutTransactions);
 
-    for (const bankTx of allTransactions) {
+    for (const transaction of allTransactions) {
       try {
-        await this.create(bankTx);
+        const bankTx = await this.create(transaction);
+        if (!DisabledProcess(Process.CREATE_TRANSACTION))
+          await this.transactionService.create({ sourceId: bankTx.id, sourceType: TransactionSourceType.BANK_TX });
```
Comment (davidleomay) [inline src/subdomains/supporting/bank-tx/bank-tx/bank-tx.service.ts:106]: Save transaction to bank TX?

File: src/subdomains/supporting/payment/entities/transaction.entity.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+  refReward: RefReward;
+
+  @OneToOne(() => BankTx, (bankTx) => bankTx.transaction, { eager: false, nullable: true })
+  bankTx: BankTx;
+
+  @OneToOne(() => CryptoInput, (cryptoInput) => cryptoInput.transaction, { eager: false, nullable: true })
+  cryptoInput: CryptoInput;
+
+  @OneToOne(() => CheckoutTx, (checkoutTx) => checkoutTx.transaction, { eager: false, nullable: true })
+  checkoutTx: CheckoutTx;
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/entities/transaction.entity.ts:60]: `eager: false` is the default

File: src/subdomains/core/referral/reward/ref-reward.service.ts
Comment (davidleomay) [inline src/subdomains/core/referral/reward/ref-reward.service.ts:1]: We need to create a transaction when creating a ref reward?

File: src/subdomains/supporting/payment/services/transaction-job.service.ts [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    private readonly payInService: PayInService,
+    private readonly checkoutTxService: CheckoutTxService,
+    private readonly transactionService: TransactionService,
+    private readonly refRewardService: RefRewardService,
+  ) {}
+
+  // --- CHECK BUY FIAT --- //
+  @Cron(CronExpression.EVERY_30_MINUTES)
+  @Lock(7200)
+  async addFiatOutputs(): Promise<void> {
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/transaction-job.service.ts:33]: Copy&paste? 😉

File: src/subdomains/supporting/payment/services/transaction-job.service.ts [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        ...(await this.refRewardService.getRewardsWithoutTransaction()),
+      ],
+      'created',
+    );
+
+    for (const unassignedTx of sortedUnassignedTx) {
+      await this.transactionService.create({
+        sourceType: this.getSourceType(unassignedTx),
+        ...this.getTypeAndEntity(unassignedTx),
+      });
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/transaction-job.service.ts:50]: Are you sure this updates the entities?

File: src/subdomains/core/buy-crypto/process/services/buy-crypto.service.ts [hunk 11]
```diff
@@ -130,13 +147,20 @@ export class BuyCryptoService {
   }
 
   async createFromCryptoInput(cryptoInput: CryptoInput, cryptoRoute: CryptoRoute): Promise<void> {
+    const transaction = !DisabledProcess(Process.CREATE_TRANSACTION)
+      ? await this.transactionService.update(cryptoInput.transaction.id, {
+          type: TransactionTypeInternal.BUY_CRYPTO,
+        })
```
Comment (davidleomay) [inline src/subdomains/core/buy-crypto/process/services/buy-crypto.service.ts:153]: `TransactionTypeInternal.CRYPTO_CRYPTO`?

File: src/subdomains/supporting/payin/entities/crypto-input.entity.ts [hunk 12]
```diff
@@ -98,6 +101,16 @@ export class CryptoInput extends IEntity {
...
 
+  @OneToOne(() => Transaction, { nullable: true })
+  @JoinColumn()
+  transaction: Transaction;
+
+  @OneToOne(() => BuyFiat, (buyFiat) => buyFiat.cryptoInput, { eager: false, nullable: true })
+  buyFiat: BuyFiat;
+
+  @OneToOne(() => BuyCrypto, (buyCrypto) => buyCrypto.cryptoInput, { eager: false, nullable: true })
+  buyCrypto: BuyCrypto;
```
Comment (davidleomay) [inline src/subdomains/supporting/payin/entities/crypto-input.entity.ts:112]: `eager: false` is default

File: src/subdomains/supporting/payin/strategies/register/impl/base/register.strategy.ts [hunk 13]
```diff
@@ -39,7 +44,13 @@ export abstract class RegisterStrategy implements OnModuleInit, OnModuleDestroy
   protected async createPayInsAndSave(transactions: PayInEntry[], log: PayInInputLog): Promise<void> {
     const payIns = transactions.map((t) => this.payInFactory.createFromEntry(t));
 
-    await this.payInRepository.saveMany(payIns);
+    for (const payIn of payIns) {
+      await this.payInRepository.save(payIn);
+      if (!DisabledProcess(Process.CREATE_TRANSACTION))
+        payIn.transaction = await this.transactionService.create({
+          sourceType: TransactionSourceType.CRYPTO_INPUT,
+        });
```
Comment (davidleomay) [inline src/subdomains/supporting/payin/strategies/register/impl/base/register.strategy.ts:52]: Add the transaction before saving? 🤔

**Review by davidleomay — APPROVED** [review]

**Pull request merged**